modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/LimitNode.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.ignite.internal.processors.query.calcite.exec.ExecutionContext;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ *
+ */
+public class LimitNode<Row> extends AbstractNode<Row> implements SingleNode<Row>, Downstream<Row> {
+    /** */
+    private static final int NOT_READY = -1;
+
+    /** */
+    private static final int LIMIT_NOT_SET = -2;
+
+    /** */
+    private final Supplier<Integer> offsetSup;
+
+    /** */
+    private final Supplier<Integer> limitSup;
+
+    /** */
+    private int offset = NOT_READY;
+
+    /** */
+    private int limit = NOT_READY;
+
+    /** */
+    private int requested;
+
+    /** */
+    private int rowNum;
+
+    /** */
+    private int waiting;
+
+    /** */
+    boolean ended;
+
+    /**
+     * @param ctx Execution context.
+     * @param offsetSup Offset parameter supplier.
+     * @param limitSup Limit parameter supplier.
+     */
+    public LimitNode(
+        ExecutionContext<Row> ctx,
+        Supplier<Integer> offsetSup,
+        Supplier<Integer> limitSup) {
+        super(ctx);
+
+        this.offsetSup = offsetSup;
+        this.limitSup = limitSup;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void request(int rowsCnt) {
+        checkThread();
+
+        assert !F.isEmpty(sources) && sources.size() == 1;
+        assert rowsCnt > 0 && requested == 0;
+
+        requested = rowsCnt;
+
+        // Initialize offset / limit.
+        if (offset == NOT_READY && limit == NOT_READY) {
+            if (offsetSup != null) {
+                offset = offsetSup.get();
+
+                if (offset < 0)
+                    onError(new IgniteSQLException("Invalid query offset: " + offset));
+            }
+            else
+                offset = 0;
+
+            if (limitSup != null) {
+                limit = limitSup.get();
+
+                if (limit < 0)
+                    onError(new IgniteSQLException("Invalid query limit: " + limit));
+            }
+            else
+                limit = LIMIT_NOT_SET;
+        }
+
+        request0(requested);
+    }
+
+    /**
+     * Process request (some parameters may not yet be calculated).
+     */
+    private void request0(int rowsCnt) {
+        if (limit == 0 || limit > 0 && rowNum >= limit + offset) {
+            end();
+
+            return;
+        }
+
+        int req = rowsCnt;
+
+        if (rowNum == 0 && offset > 0)
+            req += offset;
+
+        if (limit > 0 && rowNum + req > limit + offset)
+            req = limit + offset - rowNum;
+
+        waiting = req;
+
+        F.first(sources).request(req);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void push(Row row) {
+        checkThread();
+
+        assert downstream != null;
+
+        waiting--;
+
+        if (!ended && rowNum >= offset) {
+            requested--;
+
+            downstream.push(row);
+        }
+
+        rowNum++;
+
+        if (requested > 0 && limit > 0 && rowNum >= limit + offset)
+            end();
```

Comment:
let's cancel upstream only here and do not do it inside end() method

Reply:
Good point. Fixed